core/meta/src/TFunction.cxx

```diff
@@ -294,19 +305,23 @@ Bool_t TFunction::Update(MethodInfo_t *info)
       return kTRUE;
    } else {
       if (fInfo) {
-         R__LOCKGUARD(gInterpreterMutex);
-         gCling->MethodInfo_Delete(fInfo);
+        // The next call locks the interpreter mutex.
+        gCling->MethodInfo_Delete(fInfo);
       }
       fInfo = info;
+      // The next call locks the interpreter mutex.
```

Comment:
Since this function will (currently) take the lock up to 3 times, it might be more efficient to keep the lock for the whole section. i.e. adding a lock guard between line 306 and 307 (right after the else).

Reply:
From the comment above, you can factor out the `MethodInfo_Delete` discussion because probably that function doesn't even need locking at all, still take `GetMangledName` or `Title` as examples. And I believe the question can be generalised to all `metacling` interfaces

Reply:
> If in the end we prefer putting less locks overall, that means:

There is  distinctions between 'less lock' and 'less lock taking'.

For example, if `TFunction::Update` does not take the lock the behavior is (ignoring that we might remove the lock from delete):
```
* start Update
* * Call first delete
* * * Request the lock
* * * * * * potentially wait until it is free
* * * get lock, do some stuff, release it
* * Call second delete
* * * Request the lock
* * * * * * potentially wait until it is free
* * * get lock, do some stuff, release it
* * Call GetMangledName
* * * Request the lock
* * * * * * potentially wait until it is free
* * * get lock, do some stuff, release it
* * Call MethodInfo_Title
* * * Request the lock
* * * * * * potentially wait until it is free
* * * get lock, do some stuff, release it
```
Where in high contention mode, there might a lot of waiting and fighting over locks.
What I am proposing result in:
```
* start Update
* * Request the lock
* * * * * potentially wait until it is free
* * get lock, keep it 
* * Call first delete
* * * Request the lock .. which we already have 
* * * do some stuff
* * Call second delete
* * * Request the lock .. which we already have 
* * * do some stuff
* * Call GetMangledName
* * * Request the lock .. which we already have 
* * * do some stuff
* * Call MethodInfo_Title
* * * Request the lock .. which we already have 
* * * do some stuff
```

where we have the "best of both world":   `Update` is only waiting for the lock once (instead of 4 times) and the lower level routine are still protected (and allowing for finer grained lock when it makes sense).

it is also safer.  Adding the lock in Update is here "just" an "optimization".  Forgetting it in the case where the metacling function never things is would result in wrong behavior.

Reply:
Thank you Philippe, I understand now better your idea and I believe it would be very nice to implement this behaviour! I have a doubt though, won't we incur in deadlocks? `gInterpreterMutex` is not recursive and `TLockGuard` is not reentrant :thinking:  I'll try things out and see what happens

Reply:
After a more careful reading of the code, I see that `gInterpreterMutex` points to `ROOT::gCoreMutex` which is of type `TRWMutexImp` which itself is reentrant, so deadlocks should be avoided when locking. Can you confirm?

Reply:
Yes there is no risk of dead lock since they all try to take the same reentrant lock.